src/web/web-service/wpan_service.cpp

```diff
@@ -66,20 +70,37 @@ std::string WpanService::HandleJoinNetworkRequest(const std::string &aJoinReques
     VerifyOrExit(client.Connect(), ret = kWpanStatus_SetFailed);
 
     VerifyOrExit(reader.parse(aJoinRequest.c_str(), root) == true, ret = kWpanStatus_ParseRequestFailed);
-    index        = root["index"].asUInt();
-    masterKey    = root["masterKey"].asString();
-    prefix       = root["prefix"].asString();
-    defaultRoute = root["defaultRoute"].asBool();
+    index          = root["index"].asUInt();
+    credentialType = root["credentialType"].asString();
+    masterKey      = root["masterKey"].asString();
+    pskd           = root["pskd"].asString();
+    prefix         = root["prefix"].asString();
+    defaultRoute   = root["defaultRoute"].asBool();
 
     if (prefix.find('/') == std::string::npos)
     {
         prefix += "/64";
     }
 
     VerifyOrExit(client.FactoryReset(), ret = kWpanStatus_LeaveFailed);
-    VerifyOrExit((ret = commitActiveDataset(client, masterKey, mNetworks[index].mNetworkName, mNetworks[index].mChannel,
-                                            mNetworks[index].mExtPanId, mNetworks[index].mPanId)) == kWpanStatus_Ok);
-    VerifyOrExit(client.Execute("ifconfig up") != nullptr, ret = kWpanStatus_JoinFailed);
+
+    if (credentialType == CREDENTIAL_TYPE_MASTER_KEY)
+    {
+        VerifyOrExit(
+            (ret = commitActiveDataset(client, masterKey, mNetworks[index].mNetworkName, mNetworks[index].mChannel,
+                                       mNetworks[index].mExtPanId, mNetworks[index].mPanId)) == kWpanStatus_Ok);
+        VerifyOrExit(client.Execute("ifconfig up") != nullptr, ret = kWpanStatus_JoinFailed);
+    }
+    else if (credentialType == CREDENTIAL_TYPE_PSKD)
+    {
+        VerifyOrExit(client.Execute("ifconfig up") != nullptr, ret = kWpanStatus_JoinFailed);
+        VerifyOrExit(client.Execute("joiner start %s", pskd.c_str()) != nullptr, ret = kWpanStatus_JoinFailed);
```

Comment:
Do we need to wait until the joiner operation completes? I don't know if calling `thread start` before the joiner operation completes will produce the intended result.

Reply:
That works during my experiment. 

Currently the "Execute()" method checks whether OpenThread returns "Done" for a command then returns. Are you suggesting checking the "Join success" message?

Reply:
I think it would be safer to wait for the "Join success" message.

@bukepo , thoughts?

Reply:
Agree to check the joiner success message. And better to tell user what error is and how to fix. For example, the not found error should be caused by the joiner is not added by commissioner, so the user should use a commissioner to add it. For security error, it probably indicates the pskd does not match the one given to the commissioner, and we should ask the user double check the pskd. Ideally we generate the standard qrcode and show it in the dialogue when user types the pskd so that users can scan it with an external commissioner app.

Reply:
This is a qrcode generator supports embedded `<img>`: http://goqr.me/

Reply:
Sounds good. Thanks for the suggestions. I've added the error message for the 2 different cases and embedded the QR code generator.